apps/src/gamelab/animationListModule.js

```diff
@@ -0,0 +1,557 @@
+      return Object.assign({}, state, {
+        loadedFromSource: true,
+        saved: true,
+        blob: action.blob,
+        dataURI: action.dataURI
+      });
+
+    case ON_ANIMATION_SAVED:
+      return Object.assign({}, state, {
+        saved: true,
+        version: action.version
+      });
+
+    default:
+      return state;
+  }
+}
+
+/**
+ *
+ * @param {!SerializedAnimationList} serializedAnimationList
+ * @returns {function()}
+ */
+export function setInitialAnimationList(serializedAnimationList) {
+  return dispatch => {
+    // TODO: Should I validate here?
+    // TODO: Should I add transient properties here?
+    dispatch({
+      type: SET_INITIAL_ANIMATION_LIST,
+      animationList: serializedAnimationList
+    });
+    serializedAnimationList.list.forEach(key => {
+      dispatch(loadAnimationFromSource(key));
+    });
+  };
+}
+
+/**
+ * Add an animation to the project (at the end of the list).
+ * @param {!AnimationKey} key
+ * @param {!SerializedAnimation} data
+ */
+export function addAnimation(key, data) {
+  // TODO: Validate that key is not already in use?
+  // TODO: Validate data format?
+  return dispatch => {
+    dispatch({
+      type: ADD_ANIMATION,
+      key,
+      data
+    });
+    dispatch(loadAnimationFromSource(key, undefined, () => {
+      dispatch(selectAnimation(key));
+    }));
+    dashboard.project.projectChanged();
+  };
+}
+
+/**
+ * Add a library animation to the project.
+ * @param {!SerializedAnimation} data
+ */
+export function addLibraryAnimation(data) {
+  return dispatch => {
+    const key = utils.createUuid();
+    dispatch({
+      type: ADD_ANIMATION,
+      key,
+      data
+    });
+    dispatch(loadAnimationFromSource(key, data.sourceUrl, () => {
+      dispatch(invalidateAnimation(key));
+      dispatch(selectAnimation(key));
+    }));
+    dashboard.project.projectChanged();
+  };
+}
+
+/**
+ * Mark an animation as needing to be saved to S3.
+ * @param {AnimationKey} key
+ * @returns {{type: ActionType, key: AnimationKey}}
+ */
+function invalidateAnimation(key) {
+  return {
+    type: INVALIDATE_ANIMATION,
+    key
+  };
+}
+
+/**
+ * Clone the requested animation, putting the new one directly after the original
+ * in the project animation list.
+ * @param {!AnimationKey} key
+ * @returns {Function}
+ */
+export function cloneAnimation(key) {
+  return (dispatch, getState) => {
+    const state = getState().animationList;
+    // Track down the source animation and its index in the collection
+    const sourceIndex = state.list.indexOf(key);
+    if (sourceIndex < 0) {
+      throw new Error(`Animation ${key} not found`);
+    }
+
+    const sourceAnimation = state.data[key];
+    const newAnimationKey = utils.createUuid();
+    dispatch({
+      type: ADD_ANIMATION_AT,
+      index: sourceIndex + 1,
+      key: newAnimationKey,
+      data: Object.assign({}, sourceAnimation, {
+        name: sourceAnimation.name + '_copy', // TODO: better generated names
```

Comment:
I'd like to address this in my next PR while I build out the 'draw my own' option, which will also need generated names.